akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala

```diff
@@ -221,11 +218,12 @@ private[stream] object ConnectionSourceStage {
     override def preStart(): Unit = {
       setKeepGoing(true)
       role match {
-        case Inbound(conn, _, _) ⇒
+        case Inbound(conn, _, _, registerCallback) ⇒
           setHandler(bytesOut, readHandler)
           connection = conn
           getStageActor(connected).watch(connection)
           connection ! Register(self, keepOpenOnPeerClosed = true, useResumeWriting = false)
+          registerCallback()
```

Comment:
So before it was decremented already when the stage was materialized but now it's delayed until there is a connection - sounds right to me.